### PR TITLE
Implement Set and Dict operations/operators

### LINKED ISSUE #2: Set Operations
**Feature Request:** 
The addition of Python's set operations

**Who would use it?** 
Working with sets, doing things like union, symmetric difference, difference, and intersection on sets

**How would it work?**
Implement Python's standard set operators, namely `|`, `^`, `-`, and `&` and their augmented assignment versions

**Why should this be added?**
Anyone working with unordered sets of data could then perform these operations as they would expect within normal Python

### LINKED ISSUE #3: Dictionary Union Operations
**Feature Request:**
Adding the Python's recently implemented Dictionary Union operators, namely marge (`|`) and update (`|=`)

**Who would use it?**
Anyone working with dictionaries and merging or updating them

**How would it work?**
Implement Python's operators into the SafeDict type, applying the size limits per other operations

**Why should this be added?**
It would allow for a much more clean usage of common dictionary operations without spanning multiple lines

### PR DESCRIPTION
**Pull request by Infinidoge**
### Summary
Implements the operators and operations for sets and dicts, adding relevant limits

Resolves #2 
Resolves #3 

Requires #5 for relevant ast operators

#### PR Type

- [x] This PR is a code change that implements a feature request.
- [ ] This PR fixes an issue.
- [ ] This PR adds a new feature that is not an open feature request.
- [ ] This PR is not a code change (e.g. documentation, README, ...)
#### Other
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [x] If code changes were made then they have been tested.
- [ ] I have updated the documentation to reflect the changes.

### CODE REVIEW

**Comment by Infinidoge** [comment]
Some changes will need to be made to update this repo and avrae/avrae to use Python 3.9 (required for dictionary merge operator)

File: draconic/helpers.py [hunk 1]
```diff
@@ -279,6 +279,16 @@ def union(self, *s):
                 _raise_in_context(IterableTooLong, "This set is too large")
             return SafeSet(super().union(*s))
 
+        def intersection(self, *s):
+            if approx_len_of(self) + sum(approx_len_of(other) for other in s) > config.max_const_len:
+                _raise_in_context(IterableTooLong, "This set is too large")
+            return SafeSet(super().union(*s))
```
Comment (mommothazaz123) [inline draconic/helpers.py:285]: this should probably be `super.intersection`

File: draconic/helpers.py [hunk 2]
```diff
@@ -279,6 +279,16 @@ def union(self, *s):
...
+
+        def symmetric_difference(self, *s):
+            if approx_len_of(self) + sum(approx_len_of(other) for other in s) > config.max_const_len:
+                _raise_in_context(IterableTooLong, "This set is too large")
+            return SafeSet(super().union(*s))
```
Comment (mommothazaz123) [inline draconic/helpers.py:290]: this should probably be `super.symmetric_difference`

File: draconic/helpers.py (hunk 1 above)
Comment (mommothazaz123) [reply draconic/helpers.py:285]: also, intersection's size check shouldn't be the sum, as the size of the intersection of two sets is guaranteed to be equal to or less than the size of the smaller set. It should check each set individually to make sure each set individually is not too large.

**Review by mommothazaz123 — CHANGES_REQUESTED** [review]

File: draconic/helpers.py [hunk 3]
```diff
@@ -275,10 +275,46 @@ def add(self, element):
...
+                _raise_in_context(IterableTooLong, "This set is too large")
+            return SafeSet(super().difference(*s))
+
+        def symmetric_difference(self, *s):
+            if approx_len_of(self) + sum(approx_len_of(other) for other in s) > config.max_const_len:
+                _raise_in_context(IterableTooLong, "This set is too large")
+            return SafeSet(super().symmetric_difference(*s))
+
+        def intersection_update(self, *s):
+            self.update(self.intersection(*s))
```
Comment (mommothazaz123) [inline draconic/helpers.py:298]: This is not the same operation: 
```py
>>> a = {1, 2, 3}
>>> a.intersection_update({2, 3, 4})
>>> a
{2, 3}

>>> a = {1, 2, 3}
>>> a.update(a.intersection({2, 3, 4}))
>>> a
{1, 2, 3}
```

File: draconic/helpers.py [hunk 4]
```diff
@@ -275,10 +275,46 @@ def add(self, element):
...
+
+        def difference_update(self, *s):
+            self.update(self.difference(*s))
```
Comment (mommothazaz123) [inline draconic/helpers.py:301]: This is not the same operation:
```py
>>> a = {1, 2, 3}
>>> a.difference_update({2, 3, 4})
>>> a
{1}

>>> a = {1, 2, 3}
>>> a.update(a.difference({2, 3, 4}))
>>> a
{1, 2, 3}
```

**Review by mommothazaz123 — CHANGES_REQUESTED** [review]

File: draconic/helpers.py [hunk 5]
```diff
@@ -275,10 +275,46 @@ def add(self, element):
...
+
+        def symmetric_difference_update(self, s):
+            self.update(self.symmetric_difference(s))
```
Comment (mommothazaz123) [inline draconic/helpers.py:304]: This is not the same operation:
```py
>>> a = {1, 2, 3}
>>> a.symmetric_difference_update({2, 3, 4})
>>> a
{1, 4}

>>> a = {1, 2, 3}
>>> a.update(a.symmetric_difference({2, 3, 4}))
>>> a
{1, 2, 3, 4}
```

(it deleted my original comment for some reason)

**Review by mommothazaz123 — APPROVED** [review]
Get that green check on test coverage and LGTM 👍

**Comment by zhudotexe** [comment]
Implemented in d7c2eacad011d87473a8944671c987bcb70b0d68; closing this PR as the upstream branch was deleted which made the merge conflict annoying to resolve

**Pull request closed without merging**